allennlp/data/iterators/bucket_iterator.py

```diff
@@ -56,35 +56,40 @@ def __init__(self,
                  sorting_keys: List[Tuple[str, str]] = None,
                  padding_noise: float = 0.1,
                  biggest_batch_first: bool = False,
-                 batch_size: int = 32) -> None:
+                 batch_size: int = 32,
+                 instances_per_epoch: int = None,
+                 max_instances_in_memory: int = None) -> None:
         self._sorting_keys = sorting_keys or []
         self._padding_noise = padding_noise
         self._biggest_batch_first = biggest_batch_first
-        super(BucketIterator, self).__init__(batch_size)
+        super(BucketIterator, self).__init__(batch_size=batch_size,
+                                             instances_per_epoch=instances_per_epoch,
+                                             max_instances_in_memory=max_instances_in_memory)
 
     @overrides
     def _create_batches(self, instances: Iterable[Instance], shuffle: bool) -> Iterable[Batch]:
-        instances = ensure_list(instances)
+        for instance_list in self._memory_sized_lists(instances):
+            if self._sorting_keys:
+                instance_list = self._sort_by_padding(instance_list,
+                                                      self._sorting_keys,
+                                                      self._padding_noise)
 
-        if self._sorting_keys:
-            instances = self._sort_by_padding(instances,
-                                              self._sorting_keys,
-                                              self._padding_noise)
-        grouped_instances = list(super(BucketIterator, self)._create_batches(instances, shuffle=False))
-        if self._biggest_batch_first:
-            # We'll actually pop the last _two_ batches, because the last one might not be full.
-            last_batch = grouped_instances.pop()
-            penultimate_batch = grouped_instances.pop()
-        if shuffle:
-            random.shuffle(grouped_instances)
-        else:
-            logger.warning("shuffle parameter is set to False,"
-                           " while bucket iterators by definition change the order of your data.")
-        if self._biggest_batch_first:
-            grouped_instances.insert(0, penultimate_batch)
-            grouped_instances.insert(0, last_batch)
+            grouped_instances = list(super(BucketIterator, self)._create_batches(instance_list, shuffle=False))
+            move_to_front = self._biggest_batch_first and len(grouped_instances) > 1
+            if move_to_front:
+                # We'll actually pop the last _two_ batches, because the last one might not be full.
+                last_batch = grouped_instances.pop()
+                penultimate_batch = grouped_instances.pop() if grouped_instances else None
```

Comment:
Can you update the docstring to state that `biggest_batch_first` now uses the biggest batch per block of instances that you have specified fit in memory.